packages/ui/components/VisualEditor.tsx

```diff
@@ -0,0 +1,243 @@
+      onToggleRequired(name);
+    }
+  };
+
+  return (
+    <div className="schema-property" style={{ display: 'flex', alignItems: 'center', justifyContent: 'space-between' }}>
+      <span>
+        <strong>{name}</strong>: <span>{typeDisplay}</span>
+      </span>
+      <div>
+        <button onClick={() => onRemove(name)}>Remove</button>
+        <div
+          onClick={() => onToggleRequired(name)}
+          onKeyDown={handleKeyPress}
+          style={{
+            width: '20px',
+            height: '20px',
+            borderRadius: '50%',
+            backgroundColor: isRequired ? 'red' : 'grey',
+            cursor: 'pointer',
+            marginLeft: '10px',
+          }}
+          tabIndex={0}
+        />
+      </div>
+    </div>
+  );
+};
+
+interface SchemaObjectProps {
+    schema: Schema;
+    path: string;
+    onAdd: (path: string, updatedSchema: Schema) => void;
+    onRemove: (key: string) => void;
+    onToggleRequired: (key: string) => void;
+  }
+
+  const SchemaObject: React.FC<SchemaObjectProps> = ({ schema, path, onAdd, onRemove, onToggleRequired }) => {
+    const properties = schema.properties ?? {};
+    const propertyNames = Object.keys(properties);
+    const requiredFields = schema.required ?? [];
+
+  const handleAddProperty = (key: string, types: string[]) => {
+    const updatedSchema = {
+      ...schema,
+      properties: {
+        ...properties,
+        [key]: { type: types },
+      },
+    };
+    onAdd(path, updatedSchema);
+  };
+
+  const handleRemoveProperty = (key: string) => {
+    if (properties.hasOwnProperty(key)) {
+      const { [key]: removedProperty, ...remainingProperties } = properties;
+      const updatedRequired = requiredFields.includes(key) ? requiredFields.filter((k) => k !== key) : requiredFields;
+      const updatedSchema = {
+        ...schema,
+        properties: remainingProperties,
+        required: updatedRequired,
+      };
+      onAdd(path, updatedSchema);
+    }
+  };
+  
+  const handleToggleRequired = (key: string) => {
+    const isRequired = requiredFields.includes(key);
+    const updatedRequired = isRequired
+      ? requiredFields.filter((k) => k !== key)
+      : [...requiredFields, key];
+
+    const updatedSchema = {
+      ...schema,
+      required: updatedRequired,
+    };
+    onAdd(path, updatedSchema);
+  };
+
+  return (
+    <div className="schema-object">
+      {propertyNames.map((key) => (
+        <SchemaProperty
+          key={key}
+          name={key}
+          schema={properties[key]}
+          onRemove={handleRemoveProperty}
+          onToggleRequired={handleToggleRequired}
+          isRequired={requiredFields.includes(key)}
+        />
+      ))}
+      <PropertyControls onAdd={handleAddProperty} />
+    </div>
+  );
+};
+
+interface VisualEditorProps {
+  schema: string;
+  onSchemaChange: (updatedSchema: string) => void;
+}
+
+export const VisualEditor: React.FC<VisualEditorProps> = ({ schema, onSchemaChange }) => {
```

Comment:
Same as with CodeEditor component:

```suggestion
interface VisualEditorProps {
  schema: SchemaObject;
  onChange: (updatedSchema: SchemaObject) => void;
}

export const VisualEditor: React.FC<VisualEditorProps> = ({ schema, onChange }) => {
```